### PR TITLE
Log Page refactored to use TypeScript

### PR DESCRIPTION
**Pull request by cnasc**
# Description

This PR updates the Log page to use TypeScript throughout in anticipation of future enhancements. This is a replacement for #2107, which got out of sync with the main repo.

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: unlock-app/src/__tests__/config.test.js [hunk 1]
```diff
@@ -146,7 +146,7 @@ describe('config', () => {
     it('should contain the right URLs for chain explorers', () => {
       expect.assertions(2)
       expect(Object.keys(config.chainExplorerUrlBuilders)).toHaveLength(1)
-      expect(config.chainExplorerUrlBuilders.etherScan('0x0')).toEqual(false)
+      expect(config.chainExplorerUrlBuilders.etherScan('0x0')).toEqual('')
```
Comment (cellog) [inline unlock-app/src/__tests__/config.test.js]: let's use `toBe`

File: unlock-app/src/components/content/LogContent.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import * as UnlockTypes from '../../unlock'
+
+interface Props {
+  account: UnlockTypes.Account
+  network: UnlockTypes.Network
+  transactionFeed: UnlockTypes.Transaction[]
+  transactionMetadataMap: UnlockTypes.TransactionMetadataMap
+}
+
+export const LogContent = (props: Props) => {
```
Comment (cellog) [inline unlock-app/src/components/content/LogContent.tsx]: can we destructure in the function declaration? (don't need to change this, just wondering about the complexity level of that)

File: unlock-app/src/__tests__/components/creator/CreatorLog.test.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+afterEach(rtl.cleanup)
+describe('CreatorLog', () => {
+  beforeEach(() => {
+    wrapper = rtl.render(
+      <CreatorLog
+        transactionFeed={transactionFeed}
+        transactionMetadata={transactionMetadata} />
+    )
+  })
+  it('should show the blockNumber of the transaction', () => {
```
Comment (julien51) [inline unlock-app/src/__tests__/components/creator/CreatorLog.test.tsx]: nit: space above

File: unlock-app/src/components/content/LogContent.tsx (hunk 2 above)
Comment (julien51) [reply unlock-app/src/components/content/LogContent.tsx]: it becomes a mess with TS ;)

File: unlock-app/src/components/content/LogContent.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+              transactionMetadata={transactionMetadataMap}
+            />
+          </BrowserOnly>
+        )}
+      </Layout>
+    </GlobalErrorConsumer>
+  )
+}
+
+export const readable = (type: UnlockTypes.TransactionType) => {
```
Comment (julien51) [inline unlock-app/src/components/content/LogContent.tsx]: Could you please add a TODO to eventually use a better mechanism (string mapping!) because I am pretty sure in the long term this will come and bite us!

File: unlock-app/src/components/content/LogContent.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  let parts = type.split('_')
+  return parts.map(part => part[0] + part.slice(1).toLowerCase()).join(' ')
+}
+
+export const mapStateToProps = (
```
Comment (julien51) [inline unlock-app/src/components/content/LogContent.tsx]: Hum, this ts is gross/hard to read? can we use the same trick as we do for components? (define the interface/s above and then destructure inside the function)?

File: unlock-app/src/components/content/LogContent.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  { config: { chainExplorerUrlBuilders } } : { config: { chainExplorerUrlBuilders: UnlockTypes.ChainExplorerURLBuilders } }
+) => {
+  const transactionFeed = Object.values(transactions).sort(
+    (a, b) => b.blockNumber - a.blockNumber
+  )
+
+  const transactionMetadataMap: UnlockTypes.TransactionMetadataMap = {}
+
+  transactionFeed.forEach(tx => {
+    transactionMetadataMap[tx.hash] = {
```
Comment (julien51) [inline unlock-app/src/components/content/LogContent.tsx]: now that I read this I wonder if instead of passing `transactionMetadataMap` we should not use a react component like `<TransactionMetadata>` which would show what we need to show? That would probably be cleaner?

Let's open an issue for this and do that in a separate PR (you can get rid of the strange `TransactionMetadataMap` type too!

File: unlock-app/src/components/creator/CreatorLog.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  const { transactionFeed, transactionMetadata } = props
+  return (
+    <Grid>
+      <HeaderItem>Block Number</HeaderItem>
+      <HeaderItem>Lock Name/Address</HeaderItem>
+      <HeaderItem>Type</HeaderItem>
+      {transactionFeed.map(tx => {
+        const { href, readableName } = transactionMetadata[tx.hash]
+        return (
+          <React.Fragment key={tx.hash}>
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLog.tsx]: All of this could be a `<Transaction>` component which would display the readable name without having to pass it at state.

File: unlock-app/src/__tests__/components/content/LogContent.test.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import {
+  mapStateToProps,
+  readable,
+} from '../../../components/content/LogContent'
+import * as UnlockTypes from '../../../unlock'
+
+const { LOCK_CREATION, WITHDRAWAL, UPDATE_KEY_PRICE } = UnlockTypes.TransactionType
+const transactions: UnlockTypes.Transactions = {
+  '0x12345678': {
```
[the remark this answers is not in the document]
Comment (cnasc) [reply unlock-app/src/__tests__/components/content/LogContent.test.ts]: TypeScript doesn't do dependent types, that's going to have to be a runtime check

File: unlock-app/src/components/content/LogContent.tsx (hunk 2 above)
Comment (cnasc) [reply unlock-app/src/components/content/LogContent.tsx]: Haha yup, you can destructure, but the type declaration gets very ugly.

File: unlock-app/src/components/content/LogContent.tsx (hunk 5 above)
Comment (cnasc) [reply unlock-app/src/components/content/LogContent.tsx]: Yes, we can definitely do that. I'll make that change.

File: unlock-app/src/components/content/LogContent.tsx (hunk 2 above)
Comment (cnasc) [reply unlock-app/src/components/content/LogContent.tsx]: Oooh, did you mean can we do
```
({ account, network, transactionFeed, transactionMetadataMap}: Props) => ...
```

🤔 I didn't try, but I'm going to now.

File: unlock-app/src/components/content/LogContent.tsx (hunk 4 above)
Comment (cnasc) [reply unlock-app/src/components/content/LogContent.tsx]: My preferred solution for this would be to just change the enum values to the readable forms. Insofar as I'm aware, we never actually use those bare string values for anything but this.

File: unlock-app/src/components/content/LogContent.tsx (hunk 6 above)
Comment (cnasc) [reply unlock-app/src/components/content/LogContent.tsx]: I would just as soon get rid of the concept of transaction metadata altogether. I think the right solution for the readable type name is to simply change the enum (as mentioned above). Assuming that is done, we're just trying to pass one additional piece of information, which I think we can do in a very lightweight manner.

File: unlock-app/src/components/creator/CreatorLog.tsx (hunk 7 above)
Comment (cnasc) [reply unlock-app/src/components/creator/CreatorLog.tsx]: Can you elaborate a bit on this?

**Comment by cnasc** [comment]
Closing in favor of #2279
